File: courses/programming-languages/clojure-language-4/groupby.py
def inc(cnt, key):
    if key in cnt:
        cnt[key] += 1
    else:
        cnt[key] = 1
    return cnt

def repeat(n, x):
    return [x for i in range(n)]

def count_seq0(seq):
    if len(seq) == 0:
        return dict()
    else:
        head = seq[0]
        tail = seq[1:]
        return inc(count_seq0(tail), head)

def count_seq(seq, cnt):
    if len(seq) == 0:
        return cnt
    else:
        head = seq[0]
        tail = seq[1:]
        return count_seq(tail, inc(cnt, head))

N = 997
try:
    print count_seq0(repeat(N, "a"))
except:
    print "count_seq0 failed"

try:
    print count_seq(repeat(N, "b"), {})
except:
    print "count_seq failed"

File: courses/programming-languages/clojure-language-4/groupby.clj
(defn inc-or-1 [n] (if (nil? n) 1 (inc n)))

(defn count-seq0 [xs]
  (if (empty? xs)
    {}
    (let [head (first xs)
          tail (rest xs)]
      (update (count-seq0 tail) head inc-or-1))))

(defn count-seq [xs cnt]
  (if (empty? xs)
    cnt
    (let [head (first xs)
          tail (rest xs)]
      (recur tail (update cnt head inc-or-1))))) 

(defn -main []
  (let [N 1000001]
    (try 
      (println (count-seq0 (repeat N :a)))
      (catch Throwable e (println "count-seq0 failed")))
    (try (println (count-seq (repeat N :a) {})))))

(-main)

